### PR TITLE
specifications: configurable apt mirror

### PR DESCRIPTION
**Pull request by cjp256**
Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: specifications/configurable-apt-mirror.org [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+Enable users to configure mirror URL with fully supported arguments:
+
+- Modify Snapcraft to support the command-line option =--apt-mirror
+  <url>= with corresponding environment variable =SNAPCRAFT_APT_MIRROR=<url>=.
+
+- Replace =SNAPCRAFT_BUILD_ENVIRONMENT_PRIMARY_MIRROR= variable with introduced
+  =SNAPCRAFT_APT_MIRROR=.
+
+- Update APT configuration on every run of build environment, enabling the
```
Comment (sergiusens) [inline specifications/configurable-apt-mirror.org:39]: just like the `--use-lxd` case, how about we make this sticky?

File: todo.org [hunk 2]
```diff
@@ -68,3 +68,17 @@ CLOSED: [2020-06-24]
 - [X] Implement =snapcraft list-tracks=
 - [ ] Add documentation for track management to the [[https://snapcraft.io/docs/release-management][Release management]] docs.
 
+*** TODO [[file:specifications/configurable-apt-mirror.org][Configurable APT Mirror]] [0/7]
+- [ ] Add =--snapcraft-apt-mirror= command-line argument with matching
+  =SNAPCRAFT_APT_MIRROR= environment variable.
+- [ ] Replace usage of =SNAPCRAFT_BUILD_EVNIRONMENT_PRIMARY_MIRROR= with
```
Comment (sergiusens) [inline todo.org:74]:
```suggestion
- [ ] Replace usage of =SNAPCRAFT_BUILD_ENVIRONMENT_PRIMARY_MIRROR= with
```

File: todo.org [hunk 3]
```diff
@@ -68,3 +68,17 @@ CLOSED: [2020-06-24]
...
+  =SNAPCRAFT_APT_MIRROR=.
+- [ ] Remove configurable security mirror.
```
Comment (sergiusens) [inline todo.org:76]: The is no mention that this is configurable in the text

File: todo.org [hunk 4]
```diff
@@ -68,3 +68,17 @@ CLOSED: [2020-06-24]
...
+- [ ] Move apt configuration in base provider into own
+  =_setup_apt_configuration()=.
+- [ ] Implement abstract _pull_file() in base provider, allowing
```
Comment (sergiusens) [inline todo.org:79]: what is this for?

File: todo.org [hunk 5]
```diff
@@ -68,3 +68,17 @@ CLOSED: [2020-06-24]
...
+  the base provider to get a file from the target.
+- [ ] Call =_setup_apt_configuration()= on every run, not just
```
Comment (sergiusens) [inline todo.org:81]: if made sticky this should not be the case

**Review by sergiusens — CHANGES_REQUESTED** [review]
Looks pretty straightforward and clear, just a couple of comments

File: todo.org (hunk 3 above)
Comment (cjp256) [reply todo.org:76]: There's a broken attempt to make it configurable to remove (it uses SNAPCRAFT_BUILD_ENVIRONMENT_PRIMARY_MIRROR).

File: todo.org (hunk 4 above)
Comment (cjp256) [reply todo.org:79]: This was to check the apt configuration state to know if there were changes.  We can record this instead and add it to the compatibility checks.

File: todo.org (hunk 5 above)
Comment (cjp256) [reply todo.org:81]: Agreed

File: specifications/configurable-apt-mirror.org (hunk 1 above)
Comment (cjp256) [reply specifications/configurable-apt-mirror.org:39]: Sticky to all LXD project environments, or just the one?

**Comment by sergiusens** [comment]
Chris Patterson writes:

> @cjp256 commented on this pull request.
>
>
>
>> +** Currently Snapcraft supports configuring this on creation with
> +=SNAPCRAFT_BUILD_ENVIRONMENT_PRIMARY_MIRROR=<url>=
> +
> +- This environment variable is undocumented and has no known usage.
> +
> +* Implementation
> +
> +Enable users to configure mirror URL with fully supported arguments:
> +
> +- Modify Snapcraft to support the command-line option =--apt-mirror
> +  <url>= with corresponding environment variable =SNAPCRAFT_APT_MIRROR=<url>=.
> +
> +- Replace =SNAPCRAFT_BUILD_ENVIRONMENT_PRIMARY_MIRROR= variable with introduced
> +  =SNAPCRAFT_APT_MIRROR=.
> +
> +- Update APT configuration on every run of build environment, enabling the
>
> Sticky to all LXD project environments, or just the one?

Good question, I would think just the one to not accidentally leak
behavior to other projects. The env variable would make it "global".

**Comment by sergiusens** [comment]
Any updates on this?

**Comment by cjp256** [comment]
I removed the extraneous details that should satisfy the comments.  I'm ignoring the sticky bits for a future spec.

**Review by sergiusens — APPROVED** [review]

**Pull request merged**